File: backend/app/services/person_service.py
"""
Person Service - Lógica de negócio para pessoas/contatos.
Implementa validações e regras de negócio.
"""
from typing import Optional
from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.repositories.person_repository import PersonRepository
from app.schemas.person import PersonCreate, PersonUpdate, PersonResponse, PersonListResponse
from app.models.person import Person


class PersonService:
    """
    Service para lógica de negócio relacionada a pessoas.
    """

    def __init__(self, db: Session):
        self.db = db
        self.repository = PersonRepository(db)

    def get_person_by_id(self, person_id: int) -> Person:
        """
        Busca uma pessoa por ID.

        Args:
            person_id: ID da pessoa

        Returns:
            Person

        Raises:
            HTTPException: Se a pessoa não for encontrada
        """
        person = self.repository.find_by_id(person_id)
        if not person:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Pessoa não encontrada"
            )
        return person

    def list_persons(
        self,
        page: int = 1,
        page_size: int = 50,
        is_active: Optional[bool] = None,
        search: Optional[str] = None,
        organization_id: Optional[int] = None,
        owner_id: Optional[int] = None
    ) -> PersonListResponse:
        """
        Lista pessoas do sistema com paginação.

        Args:
            page: Número da página (começa em 1)
            page_size: Tamanho da página
            is_active: Filtro por status ativo
            search: Termo de busca (nome, email, telefone, cargo)
            organization_id: Filtro por organização
            owner_id: Filtro por responsável

        Returns:
            PersonListResponse com lista paginada de pessoas
        """
        # Calcula offset
        skip = (page - 1) * page_size

        # Busca pessoas
        persons = self.repository.list_all(
            skip=skip,
            limit=page_size,
            is_active=is_active,
            search=search,
            organization_id=organization_id,
            owner_id=owner_id
        )

        # Conta total
        total = self.repository.count_all(
            is_active=is_active,
            search=search,
            organization_id=organization_id,
            owner_id=owner_id
        )

        # Calcula total de páginas
        total_pages = (total + page_size - 1) // page_size

        # Converte para response schema
        persons_response = [
            PersonResponse.model_validate(person)
            for person in persons
        ]

        return PersonListResponse(
            persons=persons_response,
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages
        )

    def create_person(self, person_data: PersonCreate) -> Person:
        """
        Cria uma nova pessoa.

        Args:
            person_data: Dados da pessoa

        Returns:
            Person criado

        Raises:
            HTTPException: Se email já existir
        """
        # Valida se algum email já existe (verifica todos os campos de email fornecidos)
        emails_to_check = [
            person_data.email,
            person_data.email_commercial,
            person_data.email_personal,
            person_data.email_alternative
        ]

        for email in emails_to_check:
            if email and self.repository.exists_email(email):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Email {email} já cadastrado"
                )

        # Cria a pessoa
        person = self.repository.create(person_data)
        return person

    def update_person(self, person_id: int, person_data: PersonUpdate) -> Person:
        """
        Atualiza uma pessoa existente.

        Args:
            person_id: ID da pessoa
            person_data: Novos dados

        Returns:
            Person atualizado

        Raises:
            HTTPException: Se a pessoa não for encontrada ou se houver conflito
        """
        # Busca a pessoa
        person = self.get_person_by_id(person_id)

        # Valida emails (se foram alterados)
        emails_to_check = [
            ('email', person_data.email, person.email),
            ('email_commercial', person_data.email_commercial, person.email_commercial),
            ('email_personal', person_data.email_personal, person.email_personal),
            ('email_alternative', person_data.email_alternative, person.email_alternative),
        ]

        for field_name, new_email, current_email in emails_to_check:
            if new_email and new_email != current_email:
                if self.repository.exists_email(new_email, exclude_id=person_id):
                    raise HTTPException(
                        status_code=status.HTTP_400_BAD_REQUEST,
                        detail=f"Email {new_email} já cadastrado"
                    )

        # Atualiza a pessoa
        updated_person = self.repository.update(person, person_data)
        return updated_person

    def delete_person(self, person_id: int) -> None:
        """
        Deleta uma pessoa.

        Args:
            person_id: ID da pessoa

        Raises:
            HTTPException: Se a pessoa não for encontrada
        """
        # Busca a pessoa
        person = self.get_person_by_id(person_id)

        # Deleta (hard delete)
        self.repository.delete(person)

    def set_active_status(self, person_id: int, is_active: bool) -> Person:
        """
        Ativa ou desativa uma pessoa.

        Args:
            person_id: ID da pessoa
            is_active: Novo status

        Returns:
            Person atualizado

        Raises:
            HTTPException: Se a pessoa não for encontrada
        """
        # Busca a pessoa
        person = self.get_person_by_id(person_id)

        # Atualiza status
        updated_person = self.repository.set_active_status(person, is_active)
        return updated_person

    def get_persons_by_organization(self, organization_id: int) -> list[Person]:
        """
        Busca todas as pessoas de uma organização.

        Args:
            organization_id: ID da organização

        Returns:
            Lista de pessoas
        """
        return self.repository.find_by_organization(organization_id)

File: frontend/src/services/personService.ts
/**
 * Person Service - Serviço para gerenciamento de pessoas/contatos
 * Responsável por todas as operações relacionadas a pessoas
 */
import api from "./api";

export interface Person {
  id: number;
  name: string;
  first_name?: string;
  last_name?: string;

  // Emails
  email?: string;
  email_commercial?: string;
  email_personal?: string;
  email_alternative?: string;

  // Telefones
  phone?: string;
  phone_commercial?: string;
  phone_whatsapp?: string;
  phone_alternative?: string;

  // Profissional
  position?: string;

  // Redes sociais
  linkedin?: string;
  instagram?: string;
  facebook?: string;

  // Relacionamentos
  organization_id?: number;
  owner_id?: number;

  // Controle
  is_active: boolean;
  pipedrive_id?: number;

  created_at: string;
  updated_at: string;
}

export interface PersonListResponse {
  persons: Person[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CreatePersonRequest {
  name: string;
  first_name?: string;
  last_name?: string;

  // Emails
  email?: string;
  email_commercial?: string;
  email_personal?: string;
  email_alternative?: string;

  // Telefones
  phone?: string;
  phone_commercial?: string;
  phone_whatsapp?: string;
  phone_alternative?: string;

  // Profissional
  position?: string;
  organization_id?: number;
  owner_id?: number;

  // Redes sociais
  linkedin?: string;
  instagram?: string;
  facebook?: string;

  // Controle
  is_active?: boolean;
  pipedrive_id?: number;
}

export interface UpdatePersonRequest {
  name?: string;
  first_name?: string;
  last_name?: string;

  // Emails
  email?: string;
  email_commercial?: string;
  email_personal?: string;
  email_alternative?: string;

  // Telefones
  phone?: string;
  phone_commercial?: string;
  phone_whatsapp?: string;
  phone_alternative?: string;

  // Profissional
  position?: string;
  organization_id?: number;
  owner_id?: number;

  // Redes sociais
  linkedin?: string;
  instagram?: string;
  facebook?: string;

  // Controle
  is_active?: boolean;
  pipedrive_id?: number;
}

/**
 * Lista todas as pessoas com filtros e paginação
 */
const list = async (params?: {
  page?: number;
  page_size?: number;
  search?: string;
  is_active?: boolean;
  organization_id?: number;
  owner_id?: number;
}): Promise<PersonListResponse> => {
  const response = await api.get("/api/v1/persons", { params });
  return response.data;
};

/**
 * Busca uma pessoa por ID
 */
const getById = async (id: number): Promise<Person> => {
  const response = await api.get(`/api/v1/persons/${id}`);
  return response.data;
};

/**
 * Busca todas as pessoas de uma organização
 */
const getByOrganization = async (organizationId: number): Promise<Person[]> => {
  const response = await api.get(`/api/v1/persons/organization/${organizationId}`);
  return response.data;
};

/**
 * Cria uma nova pessoa
 */
const create = async (data: CreatePersonRequest): Promise<Person> => {
  const response = await api.post("/api/v1/persons", data);
  return response.data;
};

/**
 * Atualiza uma pessoa existente
 */
const update = async (id: number, data: UpdatePersonRequest): Promise<Person> => {
  const response = await api.put(`/api/v1/persons/${id}`, data);
  return response.data;
};

/**
 * Deleta uma pessoa
 */
const remove = async (id: number): Promise<void> => {
  await api.delete(`/api/v1/persons/${id}`);
};

/**
 * Altera o status ativo/inativo de uma pessoa
 */
const setActiveStatus = async (id: number, isActive: boolean): Promise<Person> => {
  const response = await api.patch(`/api/v1/persons/${id}/status`, null, {
    params: { is_active: isActive },
  });
  return response.data;
};

/**
 * Vincula uma pessoa a um card
 */
const linkToCard = async (cardId: number, personId: number): Promise<void> => {
  await api.post(`/api/v1/cards/${cardId}/person`, null, {
    params: { person_id: personId },
  });
};

/**
 * Desvincula uma pessoa de um card
 */
const unlinkFromCard = async (cardId: number): Promise<void> => {
  await api.delete(`/api/v1/cards/${cardId}/person`);
};

const personService = {
  list,
  getById,
  getByOrganization,
  create,
  update,
  delete: remove,
  setActiveStatus,
  linkToCard,
  unlinkFromCard,
};

export default personService;
